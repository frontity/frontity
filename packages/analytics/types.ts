import { Package, Action } from "frontity/types";
import Source from "@frontity/source/types";
import Router from "@frontity/router/types";

/**
 * Pageview payload passed to analytics `pageview` actions.
 */
export interface Pageview {
  /**
   * Link of the page (without the host part).
   *
   * @example "/2016/the-beauties-of-gullfoss"
   */
  link: string;

  /**
   * Title of the page.
   *
   * @example "The Beauties Of Gullfoss - Frontity Test"
   */
  title: string;
}

/**
 * Eventy payload passed to analytics `event` actions.
 */
export interface Event {
  /**
   * The name of the event.
   *
   * @example "Click"
   */
  name: string;

  /**
   * Collection of properties sent along with the event.
   *
   * @remarks
   * Each analytics package requires some specific properties to be included
   * here. Check out their documentation for more info.
   *
   * @example
   * ```
   * {
   *   label: "video",
   *   category: "youtube",
   * }
   * ```
   */
  payload: Record<string, any>;
}

/**
 * Base types and actions to build analytics packages for Frontity.
 */
interface Analytics extends Package {
  /**
   * Root elements exposed by this package.
   */
  roots: {
    /**
     * Analytics root element.
     */
    analytics: React.FC;
  };

  /**
   * The actions exposed by this package.
   */
  actions: {
    /**
     * Analytics namespace.
     */
    analytics: {
      /**
       * Send a pageview to all active analytics packages.
       *
       * This action takes all namespaces defined in `state.analytics.pageviews`
       * that are `true` and calls the `pageview` action of each one with the
       * specified `Pageview` object.
       *
       * @remarks
       * This action is used by the `roots.analytics` component and is not
       * meant to be called directly.
       *
       * @example
       * ```
       * actions.analytics.pageview({
       *   link: "/2016/the-beauties-of-gullfoss",
       *   title: "The Beauties Of Gullfoss",
       * });
       * ```
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<Analytics, Pageview>;

      /**
       * Send an event to all active analytics packages.
       *
       * This action takes all namespaces defined in `state.analytics.events`
       * that are `true` and calls the `event` action of each one with the
       * specified `Event` object.
       *
       * @example
       * ```
       * actions.analytics.event({
       *   name: "click",
       *   payload: {
       *     category: "video",
       *     label: "featured-media",
       *   },
       * });
       * ```
       *
       * @param event - Object of type {@link Event}.
       */
      event: Action<Analytics, Event>;
    };
  };

  /**
   * The state exposed by this package.
   */
  state: {
    /**
     * Analytics namespace.
     */
    analytics: {
      /**
       * Map of namespaces with boolean values.
       *
       * This object is used by `actions.analytics.pageview` to know which
       * analytics packages should send the pageview to their respective
       * services.
       *
       * @remarks
       * If you want to disable sending pageviews for a specific analytics
       * package, the respective namespace of that package should be set here to
       * `false`.
       *
       * @example
       * ```
       * pageviews: {
       *   googleAnalytics: false,
       *   comscoreAnalytics: true
       * }
       * ```
       */
      pageviews: Record<string, boolean>;

      /**
       * Map of namespaces with boolean values.
       *
       * This object is used by `actions.analytics.event` to know which
       * analytics packages should send the event to their respective services.
       *
       * @remarks
       * If you want to disable sending events for a specific analytics
       * package, the respective namespace of that package should be set here to
       * `false`.
       *
       * @example
       * ```
       * events: {
       *   googleAnalytics: true,
       *   comscoreAnalytics: false,
       * }
       * ```
       */
      events: Record<string, boolean>;
    };
  };
}

/**
 * Package types used internally by Analytics.
 */
export type Packages = Analytics & Router & Source;

export default Analytics;
