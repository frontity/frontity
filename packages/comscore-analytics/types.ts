import { ReactType } from "react";
import { Action, Package } from "frontity/types";
import Analytics, { Pageview } from "@frontity/analytics/types";

/**
 * Comscore's Tag Parameters.
 */
interface ComscoreTagParams {
  /**
   * Tag Type. Pre-populated with a constant value of `2`.
   *
   * @defaultValue 2
   */
  c1?: string;

  /**
   * Unique ID for each client assigned by Comscore.
   */
  c2: string;

  /**
   * Full Page URL.
   *
   * Manually populated full page URL, including query string variables.
   */
  c4?: string;

  /**
   * Genre of content.
   *
   * Alphanumeric value used for client specific custom classification.
   */
  c5?: string;

  /**
   * Package.
   *
   * Alphanumeric value for customized aggregation to reflect sections or site
   * centric advertising packages.
   */
  c6?: string;

  /**
   * Client Segment ID.
   *
   * Alphanumeric value for Publisherâ€™s own segment for the machinethe content
   * asset is being served to.
   */
  c15?: string;
}
declare global {
  /**
   * Extended with Comscore global variables.
   */
  interface Window {
    /**
     * Comscore main object.
     */
    COMSCORE: {
      /**
       * Function that sends tag params to Comscore.
       *
       * @remarks
       * It generates the following params automatically, which are sent to
       * Comscore along with the received ones:
       *
       * - `c7`: Full Page URL, including query string variables.
       * - `c8`: Page Title.
       * - `c9`: Referring URL.
       *
       * @param tagParams - Object of type {@link ComscoreTagParams}.
       */
      beacon: (tagParams: ComscoreTagParams) => void;
    };

    /**
     * Array that stores the first tag params generated right before the
     * Comscore library has loaded.
     */
    _comscore: Array<ComscoreTagParams>;
  }
}

/**
 * Comscore Analytics package for Frontity.
 */
interface ComscoreAnalytics extends Package {
  /**
   * Root elements exposed by this package.
   */
  roots: Analytics["roots"] & {
    /**
     * Comscore Analytics root element.
     */
    comscoreAnalytics: ReactType;
  };

  /**
   * The state exposed by this package.
   */
  state: Analytics["state"] & {
    /**
     * Comscore Analytics namespace.
     */
    comscoreAnalytics: {
      /**
       * Single tracking ID associated to a Comscore client.
       */
      trackingId?: string;

      /**
       * Array of tracking IDs associated to a Comscore client.
       */
      trackingIds?: string[];
    };
  };

  /**
   * The actions exposed by this package.
   */
  actions: Analytics["actions"] & {
    /**
     * Comscore Analytics namespace.
     */
    comscoreAnalytics: {
      /**
       * Send a pageview to all Comscore traking ids defined in
       * `state.comscoreAnalytics.trackingIds`.
       *
       * @remarks
       * This action is called automatically if
       * `state.analytics.pageviews.comscoreAnalytics` is set to `true`.
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<ComscoreAnalytics, Pageview>;
    };
  };
}

/**
 * Package types used internally in ComscoreAnalytics.
 */
export type Packages = ComscoreAnalytics & Analytics;

export default ComscoreAnalytics;
