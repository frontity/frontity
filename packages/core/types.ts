import { TransformOptions } from "babel-core";
import { Configuration } from "webpack";

export type Mode = "development" | "production";
export type Target = "module" | "es5" | "server";

export interface BabelConfigs {
  module: TransformOptions;
  es5: TransformOptions;
  server: TransformOptions;
}

export interface WebpackConfigs {
  module: Configuration;
  es5: Configuration;
  server: Configuration;
}

export interface FrontityConfig {
  outDir: string;
}

export interface Config {
  babel: BabelConfigs;
  webpack: WebpackConfigs;
  frontity: FrontityConfig;
}

export interface EntryPoints {
  name: string;
  path: string;
}

export type FrontityTags = {
  script?: string;
  link?: string;
  style?: string;
};

export type HeadTags = {
  link: string;
  base: string;
  bodyAttributes: string;
  htmlAttributes: string;
  meta: string;
  noscript: string;
  script: string;
  style: string;
  title: string;
};

export type Template = ({
  html,
  frontity,
  head
}: {
  html: string;
  frontity: FrontityTags;
  head: HeadTags;
}) => string;
