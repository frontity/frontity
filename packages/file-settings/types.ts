type Base = {
  match?: string[];
  mode?: string;
  state?: object;
};

export type Package = {
  name: string;
  active?: boolean;
  state?: object;
};

export type NormalizedPackage = Package & {
  active: boolean; // Default: true
  state: object; // Default: {}
};

type Mono = Base & {
  name?: string;
};

type Multi = Base & {
  name: string;
};

type Imported<T> = Base & {
  packages: (string | T)[];
};

type Normalized<T> = Base & {
  mode: string; // Default: html
  state: object; // Default: {}
  packages: T[];
};

export type MonoSettings<T = Package> = Imported<T> & Mono;

export type MultiSettings<T = Package> = Imported<T> & Multi;

export type Settings<T = Package> = MonoSettings<T> | MultiSettings<T>[];

export type NormalizedSettings<T = NormalizedPackage> = Normalized<T> & Multi;

export type Site = {
  name: string;
  mode: string;
  packages: string[];
};
