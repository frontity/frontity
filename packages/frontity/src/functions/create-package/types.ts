// Options passed to the `create-package` function.
export type Options = {
  // Name of the package.
  name?: string;
  // Namespace of the package.
  namespace?: string;
  // Path of the Frontity project.
  projectPath?: string;
  // Path where the package should be created (relative to `projectPath`).
  packagePath?: string;
  // Support for TypeScript.
  typescript?: boolean;
};
