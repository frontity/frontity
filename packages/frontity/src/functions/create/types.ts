import { EventEmitter } from "events";

// Options passed to the `create` function.
export type Options = {
  // Name of the app.
  name?: string;
  // Path where the app should be created.
  path?: string;
  // Support for TypeScript.
  typescript?: boolean;
  // Frontity packages that need to be installed.
  packages?: string[];
  // Frontity starter theme to clone.
  theme?: string;
  // Emitter used to send the process messages.
  emitter?: EventEmitter;
};

// Settings generated in `createPackageJson` function
// to populate the initial `package.json` file.
export type PackageJson = {
  name: string;
  version: string;
  private: boolean;
  description: string;
  keywords: string[];
  scripts: {
    dev: string;
    build: string;
    serve: string;
  };
  dependencies: {
    [key: string]: string;
  };
};
