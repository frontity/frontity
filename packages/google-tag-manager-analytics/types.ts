import { ReactType } from "react";
import { Action } from "frontity/types";
import Analytics, { Pageview, Event } from "@frontity/analytics/types";

declare global {
  /**
   * Extended with Google Tag Manager global variables.
   */
  interface Window {
    /**
     * GTM's data layer.
     */
    dataLayer: {
      /**
       * Function to add events to the GTM's data layer.
       */
      push: (payload: any) => number;
    };
  }
}

/**
 * Analytics package to use Google Tag Manager with Frontity.
 */
interface GoogleTagManagerAnalytics extends Analytics {
  /**
   * Root elements exposed by this package.
   */
  roots: Analytics["roots"] & {
    /**
     * Google Tag Manager Analytics root element.
     */
    googleTagManagerAnalytics: ReactType;
  };

  /**
   * The state exposed by this package.
   */
  state: Analytics["state"] & {
    /**
     * State properties for the Google Tag Manager package.
     */
    googleTagManagerAnalytics: {
      /**
       * Container ID associated to a Google Tag Manager account.
       */
      containerId?: string;

      /**
       * List of container IDs associated to a Google Tag Manager account.
       */
      containerIds?: string[];
    };
  };

  /**
   * The actions exposed by this package.
   */
  actions: Analytics["actions"] & {
    /**
     * Actions from the Google analytics package.
     */
    googleTagManagerAnalytics: {
      /**
       * Send a pageview event to all container IDs defined in
       * either `state.googleTagManagerAnalytics.containerId` or
       * `state.googleTagManagerAnalytics.containerIds`.
       *
       * @remarks
       * This action is called automatically if
       * `state.analytics.pageviews.googleTagManagerAnalytics` is set to `true`.
       *
       * @example
       * ```ts
       * // If you call the main analytics action with this params:
       * actions.analytics.pageview({
       *   link: "/2016/the-beauties-of-gullfoss",
       *   title: "The Beauties Of Gullfoss - Frontity Test",
       * });
       *
       * // The pageview is added to the GTM's data layer with this format:
       * {
       *   event: "pageview",
       *   link: "/2016/the-beauties-of-gullfoss",
       *   title: "The Beauties Of Gullfoss - Frontity Test",
       *   }
       * }
       * ```
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<GoogleTagManagerAnalytics, Pageview>;

      /**
       * Send an event to all container IDs defined in
       * either `state.googleTagManagerAnalytics.containerId` or
       * `state.googleTagManagerAnalytics.containerIds`.
       *
       * @example
       * ```ts
       * // If you call the main analytics action with this params:
       * actions.analytics.event({
       *   name: "Click",
       *   payload: {
       *     category: "video",
       *     label: "youtube",
       *   }
       * });
       *
       * // The event is added to the GTM's data layer with this format:
       * {
       *   event: "Click",
       *   payload: {
       *     category: "video",
       *     label: "youtube",
       *   }
       * }
       * ```
       *
       * @remarks
       * This action is called automatically when calling
       * `actions.analytics.event()` and
       * `state.analytics.pageviews.googleTagManagerAnalytics` is set to `true`.
       *
       * @param event - Object of type {@link Event}.
       */
      event: Action<GoogleTagManagerAnalytics, Event>;
    };
  };
}

export default GoogleTagManagerAnalytics;
