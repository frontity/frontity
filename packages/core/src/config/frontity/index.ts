import { FrontityConfig } from "../../../types";

export default (): FrontityConfig => ({
  outDir: "build"
});
