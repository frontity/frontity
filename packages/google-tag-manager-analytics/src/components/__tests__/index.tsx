/**
 * @jest-environment node
 */

import React from "react";
import TestRenderer from "react-test-renderer";
import { HelmetProvider } from "frontity";
import { State } from "frontity/types";
import { FilledContext } from "react-helmet-async";
import { Root as GoogleTagManager } from "..";
import GoogleTagManagerPkg from "../../../types";

const getState = (): State<GoogleTagManagerPkg> => ({
  analytics: {
    pageviews: { googleTagManagerAnalytics: true },
    events: {},
  },
  googleTagManagerAnalytics: {},
});

describe("GoogleTagManager", () => {
  test("works with a single container id", () => {
    const state = getState();

    state.googleTagManagerAnalytics.containerId = "GTM-XXXXXXX";

    const helmetContext = {};
    const rendered = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <GoogleTagManager state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
    expect(rendered).toMatchSnapshot();
  });

  test("works with multiple container ids", () => {
    const state = getState();

    state.googleTagManagerAnalytics.containerIds = [
      "GTM-XXXXXXX",
      "GTM-YYYYYYY",
    ];

    const helmetContext = {};
    const rendered = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <GoogleTagManager state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
    expect(rendered).toMatchSnapshot();
  });

  test("doesn't add anything if there's no container ids", () => {
    const state = getState();

    const helmetContext = {};
    const rendered = TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <GoogleTagManager state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
    expect(rendered).toMatchSnapshot();
  });
});
