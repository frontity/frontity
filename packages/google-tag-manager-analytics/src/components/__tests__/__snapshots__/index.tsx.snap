// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleTagManager doesn't add anything if there's no container ids 1`] = `""`;

exports[`GoogleTagManager doesn't add anything if there's no container ids 2`] = `null`;

exports[`GoogleTagManager works with a single container id 1`] = `"<script data-rh=\\"true\\" async=\\"true\\" src=\\"https://www.googletagmanager.com/gtm.js?id=GTM-XXXXXXX\\"></script>"`;

exports[`GoogleTagManager works with a single container id 2`] = `
<noscript>
  <iframe
    css={
      Object {
        "map": undefined,
        "name": "8s6lqv",
        "next": undefined,
        "styles": "
          display: none;
          visibility: hidden;
        ",
        "toString": [Function],
      }
    }
    height="0"
    src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
    title="GTM-XXXXXXX"
    width="0"
  />
</noscript>
`;

exports[`GoogleTagManager works with multiple container ids 1`] = `"<script data-rh=\\"true\\" async=\\"true\\" src=\\"https://www.googletagmanager.com/gtm.js?id=GTM-XXXXXXX\\"></script><script data-rh=\\"true\\" async=\\"true\\" src=\\"https://www.googletagmanager.com/gtm.js?id=GTM-YYYYYYY\\"></script>"`;

exports[`GoogleTagManager works with multiple container ids 2`] = `
Array [
  <noscript>
    <iframe
      css={
        Object {
          "map": undefined,
          "name": "8s6lqv",
          "next": undefined,
          "styles": "
          display: none;
          visibility: hidden;
        ",
          "toString": [Function],
        }
      }
      height="0"
      src="https://www.googletagmanager.com/ns.html?id=GTM-XXXXXXX"
      title="GTM-XXXXXXX"
      width="0"
    />
  </noscript>,
  <noscript>
    <iframe
      css={
        Object {
          "map": undefined,
          "name": "8s6lqv",
          "next": undefined,
          "styles": "
          display: none;
          visibility: hidden;
        ",
          "toString": [Function],
        }
      }
      height="0"
      src="https://www.googletagmanager.com/ns.html?id=GTM-YYYYYYY"
      title="GTM-YYYYYYY"
      width="0"
    />
  </noscript>,
]
`;
